### PR TITLE
Allow ILM step transition to the phase terminal step

### PR DESCRIPTION
**Pull request by andreidan**
ILM tries to honour the cached phase however, some steps are implicit (e.g. injected actions or the terminal policy/phase step) Currently ILM would throw an exception if the currently cached phase was removed:
```
step [{"phase":"warm","action":"complete","name":"complete"}] for index [index] with policy [my-policy] does not exist
```

This fixes this scenario allowing ILM to transition to the phase complete step even if the phase has been removed from the actual policy.

Part of #91749

### CODE REVIEW

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java [hunk 1]
```diff
@@ -91,7 +92,9 @@ public static void validateTransition(
 
         // Always allow moving to the terminal step or to a step that's present in the cached phase, even if it doesn't exist in the policy
         if (isNewStepCached == false
-            && (stepRegistry.stepExists(policyName, newStepKey) == false && newStepKey.equals(TerminalPolicyStep.KEY) == false)) {
+            && (stepRegistry.stepExists(policyName, newStepKey) == false
+                && newStepKey.equals(TerminalPolicyStep.KEY) == false
+                && newStepKey.equals(new Step.StepKey(lifecycleState.phase(), PhaseCompleteStep.NAME, PhaseCompleteStep.NAME)) == false)) {
```
Comment (gmarouli) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java:97]: Nit: The `new` here threw me off when I read it the first time. What do you think about adding a helper method in `org.elasticsearch.xpack.core.ilm.PhaseCompleteStep`:

- Either adding  `boolean isFinalStep(StepKey stepKey)` ,
- or to add a final step key constructor:
```
    public static StepKey stepKey(String phase) {
        return new StepKey(phase, NAME, NAME);
    }
```

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransitionTests.java [hunk 2]
```diff
@@ -590,16 +592,6 @@ public void testValidateTransitionToCachedStepMissingFromPolicy() {
...
-        Map<String, LifecycleAction> actions = new HashMap<>();
-        actions.put(SetPriorityAction.NAME, new SetPriorityAction(100));
-        Phase hotPhase = new Phase("hot", TimeValue.ZERO, actions);
-        Map<String, Phase> phases = Collections.singletonMap("hot", hotPhase);
-        LifecyclePolicy policyWithoutRollover = new LifecyclePolicy("my-policy", phases);
-        LifecyclePolicyMetadata policyMetadata = new LifecyclePolicyMetadata(policyWithoutRollover, Collections.emptyMap(), 2L, 2L);
-
-        ClusterState existingState = ClusterState.builder(ClusterState.EMPTY_STATE)
-            .metadata(Metadata.builder(Metadata.EMPTY_METADATA).put(meta, false).build())
-            .build();
```
[the remark this answers is not in the document]
Comment (gmarouli) [reply x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransitionTests.java:602]: <3

**Review by gmarouli — APPROVED** [review]
LGTM, thank for improving this, I left a minor comment.

**Comment by andreidan** [comment]
@elasticmachine update branch

**Comment by andreidan** [comment]
@elasticmachine update branch

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransitionTests.java [hunk 3]
```diff
@@ -632,6 +624,66 @@ public void testValidateTransitionToCachedStepMissingFromPolicy() {
...
 
+    public void testValidateTransitionToCachedStepWhenMissingPhaseFromPolicy() {
+        // we'll test the case when the warm phase was deleted and the next step is the phase complete one
+
+        LifecycleExecutionState.Builder executionState = LifecycleExecutionState.builder()
+            .setPhase("warm")
+            .setAction("migrate")
+            .setStep("check-migration")
+            .setPhaseDefinition("""
+                {
```
Comment (joegallo) [inline x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransitionTests.java:635]: This indentation looks less than optimal.

Comment (joegallo) [reply x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransitionTests.java:635]: #92713 and also acaa99bf8aab58b2b422a097a42c273a85d0263c

**Comment by andreidan** [comment]
@elasticmachine update branch

**Pull request merged**